python/pyarrow/scalar.pxi

```diff
@@ -16,1198 +16,704 @@
 
-cdef class FloatValue(ArrayValue):
+cdef class BinaryScalar(Scalar):
     """
-    Concrete class for float32 array elements.
+    Concrete class for binary-like scalars.
     """
 
-    def as_py(self):
+    def as_buffer(self):
         """
-        Return this value as a Python float.
+        Return a view over this value as a Buffer object.
         """
-        cdef CFloatArray* ap = <CFloatArray*> self.sp_array.get()
-        return ap.Value(self.index)
-
-
-cdef class DoubleValue(ArrayValue):
-    """
-    Concrete class for float64 array elements.
-    """
+        cdef CBinaryScalar* sp = <CBinaryScalar*> self.wrapped.get()
+        return pyarrow_wrap_buffer(sp.value) if sp.is_valid else None
 
     def as_py(self):
         """
-        Return this value as a Python float.
+        Return this value as a Python bytes.
         """
-        cdef CDoubleArray* ap = <CDoubleArray*> self.sp_array.get()
-        return ap.Value(self.index)
+        buffer = self.as_buffer()
+        if buffer is not None:
+            return self.as_buffer().to_pybytes()
+        else:
+            return None
 
 
-cdef class DecimalValue(ArrayValue):
-    """
-    Concrete class for decimal128 array elements.
-    """
+cdef class LargeBinaryScalar(BinaryScalar):
+    pass
 
-    def as_py(self):
-        """
-        Return this value as a Python Decimal.
-        """
-        cdef:
-            CDecimal128Array* ap = <CDecimal128Array*> self.sp_array.get()
-            c_string s = ap.FormatValue(self.index)
-        return _pydecimal.Decimal(s.decode('utf8'))
 
+cdef class FixedSizeBinaryScalar(BinaryScalar):
+    pass
 
-cdef class StringValue(ArrayValue):
+
+cdef class StringScalar(BinaryScalar):
     """
-    Concrete class for string (utf8) array elements.
+    Concrete class for string-like (utf8) scalars.
     """
 
     def as_py(self):
         """
-        Return this value as a Python unicode string.
+        Return this value as a Python string.
         """
-        cdef CStringArray* ap = <CStringArray*> self.sp_array.get()
-        return ap.GetString(self.index).decode('utf-8')
+        buffer = self.as_buffer()
+        if buffer is not None:
+            return frombytes(self.as_buffer().to_pybytes())
+        else:
+            return None
 
-    def as_buffer(self):
-        """
-        Return a view over this value as a Buffer object.
-        """
-        cdef:
-            CStringArray* ap = <CStringArray*> self.sp_array.get()
-            shared_ptr[CBuffer] buf
 
-        buf = SliceBuffer(ap.value_data(), ap.value_offset(self.index),
-                          ap.value_length(self.index))
-        return pyarrow_wrap_buffer(buf)
+cdef class LargeStringScalar(StringScalar):
+    pass
 
 
-cdef class LargeStringValue(ArrayValue):
+cdef class ListScalar(Scalar):
     """
-    Concrete class for large string (utf8) array elements.
+    Concrete class for list-like scalars.
     """
 
-    def as_py(self):
+    cdef array(self):
```

Comment:
do we want to expose this in some form publicly as well? (like you have the "raw" `value` for timestamps)

Reply:
Exposed as `values`